### PR TITLE
systemd: smart systemd interaction

### PR DESCRIPTION
**Pull request by kubiko**
Dues to the size of the change, this PR  has been broken into smaller chunks and separate PRs for efficient reviews.
This PR now serves as view to whole enchilada and place to run overall tests for comparison.

Smart systemd interaction tries to minimise number of expensive `daemon-reload` calls to systemd. This done by batching steps and only invoking single call at the end. It also leverages systemd API where one can query systemd if `daemon-reload` is needed at all, this is because sometimes systemd invokes `daemon-reload` internally, and we would only re-run it for no additional benefit.

Overall this improved snap install/refresh/remove actions by ~30% on reference arm64 quad core device.

List of related PRs will be updated here with it's states.
### Merged:
#11241 systemd: add NeedDaemonReload to the unit state
#11242 systemd: enable batched calls for systemd calls operation on units
[#11372](https://github.com/snapcore/snapd/pull/11372) interfaces/systemd: use batch systemd operations
#11353 systemd: do not reload system when enabling/disabling services

### Reference PRs and branches:
#11278 systemd: add new apis for service lifecycle handling
[wrappers-use-new-systemd-calls](https://github.com/kubiko/snapd/tree/wrappers-use-new-systemd-calls)
[interfaces-systemd-improvements](https://github.com/kubiko/snapd/tree/interfaces-systemd-improvements)
[overlord-hookstate](https://github.com/kubiko/snapd/tree/overlord-hookstate)
[overlord-servicestate](https://github.com/kubiko/snapd/tree/overlord-servicestate)

## For reference original text:

> Attempt to optimise way snapd interacts with systemd to improve snap install, fresh and first boot.
> 
> Some systemd operations are expensive, especially on slower devices.
> This change tries to batch calls when possible and at the same time it extensively tries to avoid calling `daemon-reload` which is especially expensive call. On modern systemd, every `enable`/`disable` call caused internal systemd reload. When possible services are batched and enabled at once, while additional `daemon-reload` is skipped, unless needed.
> This dramatically reduces number of `daemon-reload` calls during snap install time.
> There still some work to be done for have same level of improvement for snap remove scenario.
> 
> Gain wise, on slower embedded devices ( arm64 quad core A53), this change gains ~30% improvement for snap install and snap refresh. Snap remove gain is less, as it's harder to batch systemd unit removals.
> Snaps used for testing are rich with `gpio` and other plugs which results in systemd units, making snap install demanding.
> 
> On system with those two snaps, first boot improvement is over 1 minute.
> 

Thanks for helping us make a better snapd!
Have you signed the [license agreement](https://www.ubuntu.com/legal/contributors) and read the [contribution guide](https://github.com/snapcore/snapd/blob/master/CONTRIBUTING.md)?

### CODE REVIEW

File: systemd/systemd.go [hunk 1]
```diff
@@ -208,24 +213,27 @@ func MockJournalctl(f func(svcs []string, n int, follow bool) (io.ReadCloser, er
 type Systemd interface {
 	// DaemonReload reloads systemd's configuration.
 	DaemonReload() error
+	// DaemonReload reloads systemd's configuration if required
+	// adding flag is services are added(true) or removed(false), serviceNames list of services
```
Comment (mardy) [inline systemd/systemd.go:217]: s/is/if

or maybe "adding: true if services are added, false otherwise".

File: systemd/systemd.go [hunk 2]
```diff
@@ -312,15 +323,39 @@ type reporter interface {
 	Notify(string)
 }
 
+// SNAP_BOOTIMG_PART_NUM  is the number of available boot image partitions
+const MIN_UBUNTU_RELEASE = 18
+
+// Based on the release info, determine if systemd can be clever about
+// things like daemon-reload
+func smartSystemd() bool {
```
Comment (stolowski) [inline systemd/systemd.go:331]: Could we base this decision on systemd.Version() instead so that other distros can benefit from this as well?

Comment (kubiko) [reply systemd/systemd.go:331]: That was initial proposal when we discuss this some time ago. But I was told one should not rely simply on systemd version. But in principle I do agree with you, systemd version should be main indicator.

File: systemd/systemd.go (hunk 1 above)
Comment (kubiko) [reply systemd/systemd.go]: Fixed

**Comment by pedronis** [comment]
this is very large and it's unclear from the description whether the batching has correctness consequences in general or in the face of restarts/reboots.  We'll need to find a way to split it to make it landable.

**Comment by kubiko** [comment]
> this is very large and it's unclear from the description whether the batching has correctness consequences in general or in the face of restarts/reboots. We'll need to find a way to split it to make it landable.

yep, unfortunately it's large one.
I did try to avoid batching `start` command for services. But batting `enable`, `disable`, `stop` or `start` for sockets and timers should not have any effect on the service order.
Also, from my tests on UC20 it seems like `systemctl start <service 1> <service 3> <service 3>` honours the oder and starts services in the order passed.

Something can be potentially broken into smaller commits. But most changes is actually in the test code.

**Comment by anonymouse64** [comment]
> Also, from my tests on UC20 it seems like systemctl start <service 1> <service 3> <service 3> honours the oder and starts services in the order passed.

this is very surprising, we (well @bboozzoo IIRC, this was before I was on the snapd team) had to rewrite a good chunk of code in bionic-ish times to properly order services when being started since systemd upstream did not consider the ordering of services in a list of arguments to be "a single transaction" and thus not subject to the same ordering as if the services were all being started on boot for example.

Also yes this PR is too big, we really need to split it up as much, as possible, for example perhaps the NeedsDaemonReload=no feature can be added separately, etc. 

Also the quota changes are in conflict now, needs a master merge

File: systemd/systemd.go [hunk 3]
```diff
@@ -389,6 +425,41 @@ func (s *systemd) daemonReloadNoLock() error {
...
+}
+
+func (s *systemd) daemonReloadIfNeededWithLock(locked, adding bool, serviceNames ...string) error {
+	needReload := false
+	if s.smart {
+		// for _, service := range serviceNames {
+		// A status error will happen if systemd thinks that the service does
+		// not exist anymore. We force the reload in that case, as we assume
+		// that the unit exists at this point.
+		status, err := s.Status(serviceNames...)
```
Comment (bboozzoo) [inline systemd/systemd.go:439]: This is where https://github.com/snapcore/snapd/pull/11241 is used

**Comment by kubiko** [comment]
This PR is to be eventually abandoned as changes are being broken into smaller chunks and separate PRs.
This PR now serves as view to whole enchilada and place to run overall tests.
List of related PRs will be updated here with it's states

**Comment by mvo5** [comment]
Hey @kubiko - a lot of this PR has landed, I would like to close this one for now and maybe we can reopen (if needed) once it's cleaned up? How does that sound?

**Pull request closed without merging**